jungrapht-visualization/src/main/java/org/jungrapht/visualization/layout/algorithms/ForceAtlas2LayoutAlgorithm.java

```diff
@@ -0,0 +1,510 @@
+                .layoutModel(layoutModel)
+                .nodeData(frVertexData)
+                .nodeMasses(nodeMasses)
+                .nodeSizes(nodeSizes)
+                .initializer(initializer)
+                .random(random)
+                .build();
+
+        currentIteration = 0;
+    }
+
+    @Override
+    public boolean done() {
+        return currentIteration >= maxIterations;
+    }
+
+    private Point getFRData(V vertex) {
+        return frVertexData.computeIfAbsent(vertex, initializer);
+    }
+
+    /**
+     * Compute gravity force for given vertex.
+     * @param vertex
+     */
+    private void calcGravity(V vertex) {
+        Point xy = layoutModel.apply(vertex);
+        Point center = layoutModel.getCenter();
+        double r = xy.distanceSquared(center);
+        double gravity = -kg * nodeMasses.get(vertex) / r;
+
+        boolean locked = layoutModel.isLocked(vertex);
+
+        if (!locked) {
+            Point fvd = getFRData(vertex);
+            frVertexData.put(vertex, fvd.add(gravity * (xy.x - center.x), gravity * (xy.y - center.y)));
+        }
+    }
+
+    /**
+     * Compute attraction forces for given edge.
+     * @param edge
+     */
+    private void calcAttraction(Object edge) {
+        Graph<V, Object> graph = layoutModel.getGraph();
+        V vertex1 = graph.getEdgeSource(edge);
+        V vertex2 = graph.getEdgeTarget(edge);
+
+        boolean v1_locked = layoutModel.isLocked(vertex1);
+        boolean v2_locked = layoutModel.isLocked(vertex2);
+
+        if (v1_locked && v2_locked) {
+            return;
+        }
+        Point p1 = layoutModel.apply(vertex1);
+        Point p2 = layoutModel.apply(vertex2);
+        if (p1 == null || p2 == null) {
+            return;
+        }
+        double xDelta = p1.x - p2.x;
+        double yDelta = p1.y - p2.y;
+
+        double dist = Math.max(epsilon, Math.sqrt(xDelta * xDelta + yDelta * yDelta));
+        dist -= nodeSizes.get(vertex1) + nodeSizes.get(vertex2);
+
+        double force1;
+        double force2;
+
+        if (dist <= 0) {
+            force1 = force2 = 0.0;
+        } else {
+            if (useLinLog) {
+                force1 = force2 = Math.log(1 + dist);
+            } else if (dissuadeHubs) {
+                force1 = dist / nodeMasses.get(vertex1);
+                force2 = dist / nodeMasses.get(vertex2);
+            } else if (attractionByWeights) {
+                force1 = force2 = Math.pow(graph.getEdgeWeight(edge), weightsDelta) * dist;
+            } else {
+                force1 = force2 = dist;
+            }
+        }
+
+        if (Double.isNaN(force1) || Double.isNaN(force2))
+            throw new IllegalArgumentException(
+                    "Unexpected mathematical result in FRLayout:calcPositions");
+
+        if (!v1_locked) {
+            Point fvd1 = getFRData(vertex1);
+            frVertexData.put(vertex1, fvd1.add(-force1 * xDelta, -force1 * xDelta));
```

Comment:
I think this should be:
` frVertexData.put(vertex1, fvd1.add(-force1 * xDelta, -force1 * yDelta));`
(yDelta in the 2nd argument)

Reply:
I'm sorry about it. I've tested my local version and made these mistakes when move it to repo.